org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeLensProvider.java

```diff
@@ -0,0 +1,97 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.ls.core.internal.handlers;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.ISourceRange;
+import org.eclipse.jdt.core.ISourceReference;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.ResourceUtils;
+import org.eclipse.lsp4j.CodeLens;
+import org.eclipse.lsp4j.Range;
+
+/**
+ * CodeLens Provider interface. Extensions could extend codelens collection and
+ * resolution by implementing the interface
+ */
+public interface CodeLensProvider {
+	/**
+	 * Resolve codelens.
+	 *
+	 * @param lens
+	 * @param monitor
+	 * @return resolved codelens
+	 */
+	CodeLens resolveCodeLens(CodeLens lens, IProgressMonitor monitor);
+
+	/**
+	 * get type of the provider.
+	 *
+	 * @return type
+	 */
+	String getType();
+
+	/**
+	 * Collect codelens.
+	 *
+	 * @param unit
+	 *            The compilation unit of a source file.
+	 * @param monitor
+	 *            monitor.
+	 * @return A list of codelens.
+	 * @throws JavaModelException
+	 */
+	List<CodeLens> collectCodeLenses(ICompilationUnit unit, IProgressMonitor monitor) throws JavaModelException;

```

Comment:
I think ICompilationUnit should be replaced with ITypeRoot, so that it works for class files too. See #421